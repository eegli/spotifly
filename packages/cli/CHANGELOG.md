# @spotifly/cli

## 3.0.1

### Patch Changes

- 19e988c: Upgrade dependencies and remove outdated packages
- Updated dependencies [19e988c]
  - @spotifly/auth-token@3.0.1
  - @spotifly/library@3.0.1
  - @spotifly/utils@2.0.1

## 3.0.0

### Major Changes

- cd73d25: The minimum node version has been bumped from 14 to 18.

### Minor Changes

- cd73d25: The core CLI now supports the `profiles` command to list available Spotifly profiles.

### Patch Changes

- Updated dependencies [cd73d25]
  - @spotifly/auth-token@3.0.0
  - @spotifly/library@3.0.0
  - @spotifly/utils@2.0.0
  - @spotifly/core@2.0.0

## 2.1.1

### Patch Changes

- f71324f: Fix: When an access token is passed as a flag, the long-term credentials in the config file are ignored and not used to override the passed token.

## 2.1.0

### Minor Changes

- a2a3cbf: Read from a (global) config to enable auto-authentication. [Docs](https://spotifly.nougat.dev/docs/command-line#configuration-and-auto-authentication).

### Patch Changes

- Updated dependencies [a2a3cbf]
  - @spotifly/core@1.2.4
  - @spotifly/library@2.0.2

## 2.0.1

### Patch Changes

- cd4910c: Fix `--help` / `h` commands not being recognized by CLI
- Updated dependencies [19ac71f]
  - @spotifly/auth-token@2.0.1
  - @spotifly/library@2.0.1
  - @spotifly/utils@1.0.1

## 2.0.0

### Major Changes

- 036f524: Command line flags are now expected in kebab-case

### Patch Changes

- Updated dependencies [6da004d]
- Updated dependencies [036f524]
  - @spotifly/auth-token@2.0.0
  - @spotifly/library@2.0.0

## 1.0.5

### Patch Changes

- Updated dependencies [a297829]
  - @spotifly/auth-token@1.0.1
  - @spotifly/library@1.1.1

## 1.0.4

### Patch Changes

- Updated dependencies [87dbccd]
  - @spotifly/library@1.1.0

## 1.0.3

### Patch Changes

- @spotifly/library@1.0.3

## 1.0.2

### Patch Changes

- @spotifly/library@1.0.2

## 1.0.1

### Patch Changes

- @spotifly/library@1.0.1

## 1.0.0

### Major Changes

- 6f3a103: All packages are 99% stable. This is enough for the initial release ðŸŽ‰

### Patch Changes

- Updated dependencies [6f3a103]
  - @spotifly/auth-token@1.0.0
  - @spotifly/library@1.0.0
  - @spotifly/utils@1.0.0
