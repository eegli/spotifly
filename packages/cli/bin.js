#!/usr/bin/env node

require('./dist/spotifly-cli.cjs').run(process.argv.slice(2));
