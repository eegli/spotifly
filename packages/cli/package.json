{
  "name": "@spotifly/cli",
  "version": "3.0.0",
  "description": "The CLI for Spotifly",
  "keywords": [
    "cli",
    "spotify",
    "spotifly",
    "music",
    "library",
    "history",
    "data"
  ],
  "homepage": "https://spotifly.nougat.dev/docs/command-line",
  "repository": {
    "type": "git",
    "url": "https://github.com/eegli/spotifly"
  },
  "license": "MIT",
  "author": {
    "name": "Eric Egli",
    "email": "lx180777+npm@gmail.com"
  },
  "main": "dist/spotifly-cli.cjs.js",
  "bin": {
    "spotifly": "bin.js"
  },
  "files": [
    "bin.js",
    "dist",
    "cli"
  ],
  "dependencies": {
    "@eegli/tinyparse": "^0.18.0",
    "@spotifly/auth-token": "workspace:^",
    "@spotifly/core": "workspace:^",
    "@spotifly/library": "workspace:^",
    "@spotifly/utils": "workspace:^",
    "ini": "^4.1.1"
  },
  "engines": {
    "node": ">=18"
  },
  "publishConfig": {
    "directory": "_release/package"
  },
  "devDependencies": {
    "@types/ini": "^4.1.0"
  }
}
