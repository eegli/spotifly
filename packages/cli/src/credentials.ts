import { AuthProvider, RefreshTokenConfig } from '@spotifly/core/provider';
import { existsSync, readFileSync } from 'fs';
import ini from 'ini';
import os from 'os';
import { join } from 'path';

const configFileName = '.spotifly';

export function profileFromArgv(argv: string[]): string {
  const withProfile = argv.indexOf('--profile');
  if (withProfile === -1) return 'default';
  return argv[withProfile + 1];
}

export function readConfig(): string | null {
  const curDirConf = join(process.cwd(), configFileName);
  if (existsSync(curDirConf)) return readFileSync(curDirConf, 'utf-8');

  // Fallback dir is home
  const homeDirConf = join(os.homedir(), configFileName);
  if (existsSync(homeDirConf)) return readFileSync(homeDirConf, 'utf-8');

  return null;
}

export function credentialsFromConfig(
  config: string,
  profile: string
): RefreshTokenConfig {
  const parsedConfig = ini.parse(config);
  if (!parsedConfig[profile]) {
    throw new Error(`Profile "${profile}" not found in config`);
  }
  const { spt_client_id, spt_client_secret, spt_refresh_token } =
    parsedConfig[profile];
  if (!spt_client_id || !spt_client_secret || !spt_refresh_token) {
    throw new Error('Missing Spotify credentials');
  }
  return {
    clientId: spt_client_id,
    clientSecret: spt_client_secret,
    refreshToken: spt_refresh_token,
  };
}

export async function getAccessToken(credentials: RefreshTokenConfig) {
  return (await AuthProvider.getAccessToken(credentials)).access_token;
}
