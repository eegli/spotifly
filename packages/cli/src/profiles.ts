import { profilesFromConfig } from '@spotifly/utils/profile';

export const readProfiles = (): [null, string] | [string, null] => {
  const maybeProfiles = profilesFromConfig();
  if (typeof maybeProfiles === 'string') {
    return [null, maybeProfiles];
  }
  const [rawProfiles, configPath] = maybeProfiles;
  if (rawProfiles.length === 0) {
    return [null, 'Found no valid profiles in ' + configPath];
  }
  const profiles = rawProfiles.map(p => '* ' + p).join('\n');
  const toLog =
    'Available profiles:' +
    '\n\n' +
    profiles +
    '\n\n' +
    'Config file: ' +
    configPath;

  return [toLog, null];
};
