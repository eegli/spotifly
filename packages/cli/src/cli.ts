import * as authCli from '@spotifly/auth-token/cli';
import * as libraryCli from '@spotifly/library/cli';
import { colors } from '@spotifly/utils';
import ownPackage from '../package.json';

type Invoke = {
  callback: (args: string[]) => unknown;
  help: (title: string) => string;
  pkg: {
    name: string;
    version: string;
    homepage: string;
  };
};

export const invoke = async (
  argv: string[],
  { callback, help, pkg }: Invoke
): Promise<unknown> => {
  if (argv.length < 3) {
    console.info(`${colors.bold(colors.cyan(`${pkg.name} v${pkg.version}`))}

${help('Command-line usage')}
      
  For docs & help, visit ${pkg.homepage}
      `);
    return;
  }
  return callback(argv);
};

export const run = async (): Promise<unknown> => {
  const argv = process.argv.slice(2);
  const cmd = argv[0];

  switch (cmd) {
    case '--version':
    case '-v':
      console.info(
        `${colors.bold(
          colors.cyan(`${ownPackage.name} v${ownPackage.version}`)
        )}`
      );
      return;
    case '--help':
    case '-h':
    case undefined:
      console.info(`${colors.bold(
        colors.cyan(`${ownPackage.name} v${ownPackage.version}`)
      )}
- ${ownPackage.description}

Available commands: ${colors.green(`
- auth-token
- library 
`)}
For docs & help, visit ${ownPackage.homepage}`);
      return;
    case 'auth-token':
      return invoke(argv, {
        callback: authCli.callback,
        help: authCli.help,
        pkg: authCli.pkg,
      });
    case 'library':
      return invoke(argv, {
        callback: libraryCli.callback,
        help: libraryCli.help,
        pkg: libraryCli.pkg,
      });
    default:
      console.info(
        colors.yellow(
          `Unknown argument '${cmd}'.\nRun 'spotifly --help' for available commands`
        )
      );
  }
};
