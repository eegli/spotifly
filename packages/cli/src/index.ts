import { run } from './cli';

(async () => {
  await run();
})();
