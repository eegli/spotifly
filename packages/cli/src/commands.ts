import { colors as c } from '@spotifly/utils';
import ini from 'ini';
import pkg from '../package.json';
import { readConfigWithPath } from './credentials';

const lb = '\n';
const lblb = '\n\n';

const versionHeader = (name: string, version: string) => {
  return c.bold(c.cyan(`${name} v${version}`));
};
const helpFooter = (homepage: string) => {
  return `For docs & help, visit ${homepage}`;
};
const profiles = () => {
  try {
    const spotiflyConfigAndPath = readConfigWithPath();
    if (!spotiflyConfigAndPath) throw new Error();
    const [config, configPath] = spotiflyConfigAndPath;
    const parsedConfig = ini.parse(config);
    const profiles = Object.keys(parsedConfig)
      .map(p => '* ' + p)
      .join('\n');
    return `Available profiles:
      
${c.green(profiles)}
      
Config file: ${configPath}`;
  } catch (err) {
    return 'No profiles found, does your config file exist?';
  }
};
const fallback = (cmd: string) => {
  return (
    c.yellow(`Unknown argument '${cmd}'`) +
    lblb +
    "Run 'spotifly --help' for available commands"
  );
};
const subCommands = `Available subcommands: ${c.green(`
* auth 
* library
* profiles
`)}
Optional global flags:
${c.yellow('* --profile')} [string]
The profile in your Spotifly config file to use for authentication.
Defaults to 'default'.

- In order to get help for a specific subcommand, run ${c.green(
  'spotifly <subcommand> --help',
)}
- In order to see a list of available profiles, run ${c.green(
  'spotifly profiles',
)}`;

export default {
  packageHelp(
    packageName: string,
    packageVersion: string,
    packageSpecificHelp: () => string,
  ) {
    return (
      versionHeader(packageName, packageVersion) +
      lblb +
      packageSpecificHelp() +
      lblb +
      helpFooter(pkg.homepage)
    );
  },
  mainHelp() {
    return (
      versionHeader(pkg.name, pkg.version) +
      lb +
      `- ${pkg.description}` +
      lblb +
      subCommands +
      lblb +
      helpFooter(pkg.homepage)
    );
  },
  profiles,
  version() {
    return versionHeader(pkg.name, pkg.version);
  },
  fallback(cmd: string) {
    return fallback(cmd);
  },
};
