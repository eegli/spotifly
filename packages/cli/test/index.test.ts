import * as profileUtils from '@spotifly/utils/profile';
import * as cli from '../src/cli';
import { readProfiles } from '../src/profiles';

const consoleLogSpy = jest
  .spyOn(global.console, 'log')
  .mockImplementation(jest.fn);

const profileUtilsSpy = jest.spyOn(profileUtils, 'profilesFromConfig');

beforeEach(() => {
  jest.clearAllMocks();
});

[['--version'], ['-V']].forEach(args => {
  test('command ' + args[0], async () => {
    await cli.run(args);
    expect(consoleLogSpy).toHaveBeenCalledTimes(1);
    expect(consoleLogSpy.mock.calls[0][0]).toMatch(/\d.\d.\d/g);
  });
});

[['--help'], ['-h']].forEach(args => {
  test('command ' + args[0], async () => {
    await cli.run(args);
    expect(consoleLogSpy).toHaveBeenCalledTimes(1);
    expect(consoleLogSpy.mock.calls[0][0]).toMatchSnapshot();
  });
});

test('profiles with valid profile', async () => {
  profileUtilsSpy.mockReturnValue({
    success: true,
    value: [['profile1', 'profile2'], 'my-path'],
  });
  const result = readProfiles();
  expect(result).toMatchInlineSnapshot(`
    {
      "success": true,
      "value": "Available profiles:

    * profile1
    * profile2

    Config file: my-path",
    }
  `);
});
test('profiles with no profiles', async () => {
  profileUtilsSpy.mockReturnValue({
    success: true,
    value: [[], '/my/config/path'],
  });
  const result = readProfiles();
  expect(result).toMatchInlineSnapshot(`
    {
      "error": "Found no valid profiles in /my/config/path",
      "success": false,
    }
  `);
});
test('profiles with error', async () => {
  profileUtilsSpy.mockReturnValue({
    success: false,
    error: 'Invalid profile',
  });
  const result = readProfiles();
  expect(result).toMatchInlineSnapshot(`
    {
      "error": "Invalid profile",
      "success": false,
    }
  `);
});
