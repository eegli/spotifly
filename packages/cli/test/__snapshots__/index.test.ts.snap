// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command --help 1`] = `
"The CLI for Spotifly

Usage: spotifly [command?] <...flags>

Commands
   auth        Authenticate with Spotify
   library     Manage your Spotify library
   profiles    List available profiles
   help        Print this help message

Optional flags
   -h, --help      Print this help message
   -V, --version   Print the version"
`;

exports[`command -h 1`] = `
"The CLI for Spotifly

Usage: spotifly [command?] <...flags>

Commands
   auth        Authenticate with Spotify
   library     Manage your Spotify library
   profiles    List available profiles
   help        Print this help message

Optional flags
   -h, --help      Print this help message
   -V, --version   Print the version"
`;
